src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/ResourceColumnPredicatePushdownRewriter.cs

```diff
@@ -0,0 +1,100 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.Health.Fhir.Core.Features.Search.Expressions;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search.Expressions.Visitors
+{
+    /// <summary>
+    /// Promotes predicates applied directly in on the Resource table to the search parameter tables.
+    /// These are predicates on the ResourceSurrogateId and ResourceType columns. The idea is to make these
+    /// queries as selective as possible.
+    /// </summary>
+    internal class ResourceColumnPredicatePushdownRewriter : SqlExpressionRewriterWithInitialContext<object>
+    {
+        public static readonly ResourceColumnPredicatePushdownRewriter Instance = new ResourceColumnPredicatePushdownRewriter();
+
+        public override Expression VisitSqlRoot(SqlRootExpression expression, object context)
+        {
+            if (expression.SearchParamTableExpressions.Count == 0 || expression.ResourceTableExpressions.Count == 0 ||
+                expression.SearchParamTableExpressions.All(e => e.Kind == SearchParamTableExpressionKind.Include))
+            {
+                // if only Include expressions, the case is handled in IncludeMatchSeedRewriter
+                return expression;
+            }
```

Comment:
nit: maybe it's worth to push that check directly into `IncludeMatchSeedRewriter` as a static method?